Add sync pull remote site

sejas opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

- Related to https://github.com/Automattic/dotcom-forge/issues/9355

## Proposed Changes

- Add pull functionality to import a remote site

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Run `STUDIO_SITE_SYNC=true npm start`
- Click on the Sync tab
- Connect a site
- Click on pull
- Wait until the process finishes
- Observe your site was imported correctly
- Repeat the process and change tabs and sites, observe the process continues correctly
- To reproduce an error, you can comment [these lines](https://github.com/Automattic/studio/blob/27f6e6bcb1b7baf8035d5ec794a2a303bb396ffb/src/hooks/use-fetch-wpcom-sites.tsx#L49-L57) connect a simple site, then change click on pull.

> [!NOTE]  
> The pull feature will work in all sites where the import process works.



<!-- https://github.com/user-attachments/assets/24485f49-d8b9-4011-9b8b-78c0b2edfbbe !-->



<img width="1012" alt="Screenshot 2024-11-13 at 14 10 03" src="https://github.com/user-attachments/assets/65b69459-bdf4-4c39-9666-d0c1d186db7b">
<img width="1012" alt="Screenshot 2024-11-13 at 14 09 54" src="https://github.com/user-attachments/assets/9e58cb90-29d5-4602-a351-32bf14a4eb0c">
<img width="1012" alt="Screenshot 2024-11-13 at 15 15 19" src="https://github.com/user-attachments/assets/59ef137a-8e2e-4abf-af43-a48b2a7dbc4f">




## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

sejas committed:
Create basic logic with new context

sejas committed:
create download function to be used as in sciprts and ipc handler

sejas committed:
create states and progress to download a remote site zip

sejas committed:
Merge branch 'trunk' of github.com:Automattic/studio into add/sync-pull-remote-site

sejas committed:
remove calculated pull states type

sejas committed:
create auxiliar function onBackupCompleted

sejas committed:
pass selected site

sejas committed:
Pass the correct filetype

sejas committed:
import tar.gz jetpack backup

sejas committed:
Add provider on content tab sync tests

sejas committed:
Merge branch 'trunk' of github.com:Automattic/studio into add/sync-pull-remote-site

sejas committed:
move pull state progress info to its own hook

sejas committed:
disable importing when any site is pulling

sejas committed:
disconect pull push and disconnect when a site is pulling

sejas committed:
Display min width and correct font type for progress indicator

sejas committed:
handle error when creating backup

sejas committed:
Merge branch 'trunk' of github.com:Automattic/studio into add/sync-pull-remote-site

sejas committed:
Fix tests

katinthehatsite reviewed (commented):
I am still reviewing but wanted to say that I love this functionality and seeing it work feels like 🪄 ❤️ 

Some small things I noted in terms of the user experience so far:

* Once the pull operation is finished, the user gets the notification that the import is completed. I am guessing that we are re-using the import functionality here. Should we change the wording of the notification for this case to make it more relevant to the pull operation?
* Once the pulling operation is completed and you navigate to `Import/Export` tab, you are presented with the screen that mentions that the import was finished:

<img width="877" alt="Screenshot 2024-11-01 at 2 58 59 PM" src="https://github.com/user-attachments/assets/1eb262c0-ee28-420b-a544-a884dcd85ce2">

Since the user technically did not perform any operations on this tab, should we keep the initial state for it to avoid confusing the user?

katinthehatsite commented on src/components/content-tab-import-export.tsx:
```diff
@@ -180,7 +182,7 @@ const ImportSite = ( props: { selectedSite: SiteDetails } ) => {
 					isInitial={ isInitial }
 					openFileSelector={ openFileSelector }
-					disabled={ isSiteExporting }
+					disabled={ isSiteExporting || isAnySitePulling }
```
Should we also block the buttons to export the site on `Import/Export` tab while the pulling operation is running? It seems that I am still able to export when the pull is in progress?

Additionally, I noticed that it seems that when we are pulling, we are disabling the import on all local sites. Should we keep it enabled all other sites and disable it just on the site that is in pull operation progress?

sejas replied:
Blocking export when pulling a site: https://github.com/Automattic/studio/pull/612/commits/e8a0a07a418e9a903e2ee7a7083d680d2757ee20

I don't think we want to support importing different sites when pulling. This could be a follow-up if we consider it necessary.

katinthehatsite replied:
sounds good, that seems reasonable for now 👍

katinthehatsite commented on src/components/sync-connected-sites.tsx:
```diff
@@ -104,6 +111,7 @@ export function SyncConnectedSites( {
 								variant="link"
 								className="!ml-auto !text-a8c-gray-70 hover:!text-a8c-red-50 "
 								onClick={ () => handleDisconnectSite( section.id, section.name ) }
+								disabled={ isAnySitePulling }
```
What is the reason for disabling the `Pull` button for all Studio sites when one of the Studio site is pulling?

matt-west replied:
I disabled the buttons in the design as I assumed it would be more complex to handle multiple Push/Pull operations at the same time. Happy to lift the limitation if that's not the case though.

sejas replied:
I think it's a reasonable limitation. We could optimize it in the future, but if the user imports multiple sites at the same time we may need to lead with multiple threads, etc.

I tested the import flow to compare, and confirmed that even the UI is enabled for the rest of sites, the import flow doesn't work for the second site.

I've created an issue to disable the import UI for that case:
- https://github.com/Automattic/dotcom-forge/issues/9759

fredrikekelund commented on src/hooks/sync-sites-context.tsx:
```diff
@@ -0,0 +1,212 @@
+					throw new Error( 'Pull request failed' );
+				}
+			} catch ( error ) {
+				console.error( error );
```
This is sent to Sentry, right?

sejas replied:
Not really. I've updated the code to capture the sentry error https://github.com/Automattic/studio/pull/612/commits/fe30df27d478d65b1520ad88f663157711a37444 and output the response error: https://github.com/Automattic/studio/pull/612/commits/27a821444c833231071eecc791491fab2b8aa2cc

katinthehatsite commented on src/hooks/sync-sites-context.tsx:
```diff
@@ -0,0 +1,212 @@
+	);
+}
+
+function useSyncPull() {
```
What do you think about moving this into a separate file and then passing it through the context? I am worried that once we add push operation and possible other things here, the file would become large and hard to manage.

sejas replied:
I believe it’s a great idea. I had concerns about the temptation to use the “smaller” hooks within other components, but I found an effective way to minimize the risk of misuse. I’ve moved the state into the context and passed it to useSyncPull, making it clear that the state is managed by a higher-level hook.

https://github.com/Automattic/studio/pull/612/commits/a22422f58a3579c7c9d939ba48635a91ce44a95d

katinthehatsite commented on src/hooks/sync-sites-context.tsx:
```diff
@@ -0,0 +1,212 @@
+
+	const onBackupCompleted = useCallback(
+		async ( remoteSiteId: number, downloadUrl: string, selectedSite: SiteDetails ) => {
+			setPullStates( ( prevStates ) => ( {
```
I see that we are repeating a similar logic (besides changing the status) for a couple of times. What do you think about extracting `setPullStates` into some sort of a helper function?

sejas replied:
I see your eye is sharper than ever and you see improvements in patterns we followed in the past.
Here is the refactor: https://github.com/Automattic/studio/pull/612/commits/a8d64970d85bc8bc7b5dae0d54c0f208f9a9ee3f

Thanks for suggesting it.

katinthehatsite replied:
> I see your eye is sharper than ever  

😂

katinthehatsite commented on src/hooks/sync-sites-context.tsx:
```diff
@@ -0,0 +1,212 @@
+		};
+	}, [ pullStates, getBackup ] );
+
+	const isAnySitePulling = useMemo( () => {
```
Already mentioned this before, but what do you think about limiting the pull operations per local Studio site and not for all Studio sites that the user has?

sejas replied:
Let's limit all operations, at least for now. Feel free to create an issue if you think it will improve the UX for most users.

fredrikekelund commented on src/hooks/sync-sites-context.tsx:
```diff
@@ -0,0 +1,212 @@
+
+		return () => {
+			Object.values( intervals ).forEach( clearInterval );
+		};
```
```suggestion
		Object.entries( pullStates ).forEach( ( [ remoteSiteId, state ] ) => {
			if ( state.backupId && state.status.key === 'in-progress' ) {
				getBackup( Number( remoteSiteId ) );
			}
		} );
```
We are susceptible to race conditions here if a request takes more than 2s to finish. The beautiful thing is that because every `getBackup` call mutates `pullStates`, we don't actually need to set an interval here—we already get recursion for free 🙂

sejas replied:
Thanks for pointing out to the race condition, I changed it for setTimeout (https://github.com/Automattic/studio/pull/612/commits/33b521a0da83f11aa6c8b1d44eda7cd888cd6123) which will prevent any race condition.
I decided to add some delay between request as hundreds of users pulling production sites during hours could have some unnecessary overload in our systems, if we add a small delay it can reduce any risk and the UX will be the same. I've heard that some production sites take several hours creating the backups.

katinthehatsite commented:
Another note - should we block the `Start` button when the pull is in progress?  I noticed that we block it for part of the process when the backup is importing. I think it would make sense to block it for the whole time:

<img width="907" alt="Screenshot 2024-11-01 at 3 33 18 PM" src="https://github.com/user-attachments/assets/56495457-5d17-4735-8419-4724f05c2645">

<img width="924" alt="Screenshot 2024-11-01 at 3 32 55 PM" src="https://github.com/user-attachments/assets/1314f42a-210f-4bff-ad00-a1e7aac4807a">

fredrikekelund commented on src/hooks/sync-sites-context.tsx:
```diff
@@ -0,0 +1,212 @@
+					return;
+				}
+				onBackupCompleted( remoteSiteId, downloadUrl, selectedSite );
+			}
```
```suggestion
			if ( hasBackupCompleted && downloadUrl ) {
				// Replacing the 'in-progress' status will stop the active listening for the backup completion
				const selectedSite = pullStates[ remoteSiteId ]?.selectedSite;
				if ( ! selectedSite ) {
					return;
				}
				onBackupCompleted( remoteSiteId, downloadUrl, selectedSite );
			} else {
				setPullStates( ( prevStates ) => ( {
					...prevStates,
					[ remoteSiteId ]: {
						...prevStates[ remoteSiteId ],
						status: statusWithProgress,
						downloadUrl,
					},
				} ) );
			}
```
Nit, but this seems somewhat clearer to me, since `onBackupCompleted` will overwrite `pullStates` with new data immediately anyway.

sejas replied:
Thanks for the suggestion! https://github.com/Automattic/studio/pull/612/commits/e7387dfd984c6c0e755202addd863914814d9d4c

fredrikekelund commented on src/hooks/sync-sites-context.tsx:
```diff
@@ -0,0 +1,212 @@
+			} );
+
+			const statusWithProgress =
+				pullStatesProgressInfo[ response.status ] || pullStatesProgressInfo.failed;
```
Not a huge deal, but having one key in `pullStatesProgressInfo` called `completed` and another called `finished` is quite confusing. If we didn't map `response.status` directly to `pullStatesProgressInfo` here, then maybe we could avoid that.

fredrikekelund commented on src/hooks/use-sync-states-progress-info.ts:
```diff
@@ -0,0 +1,58 @@
+		| 'backup-sync-downloading'
+		| 'backup-sync-importing'
+		| 'backup-sync-finished'
+		| 'failed';
```
It's kind of confusing why these differ from the object keys in `pullStatesProgressInfo`.

fredrikekelund commented on src/components/sync-connected-sites.tsx:
```diff
@@ -132,15 +140,40 @@ export function SyncConnectedSites( {
+											<div className="a8c-body-small">
+												{ pullStates[ connectedSite.id ]?.status.message }
+											</div>
+											<ProgressBar
+												value={ pullStates[ connectedSite.id ]?.status.progress }
+												maxValue={ 100 }
+											/>
```
Not a super big deal, but it would be cool to inject some life into this progress bar… I think we could update the server code to extract a percentage of how far the site backup job has come.

sejas replied:
Agreed, I was not convinced with the state names. I applied a more consistent names here https://github.com/Automattic/studio/pull/612/commits/41d070d7263cd5a9e96094341d2f0bb26c85caf7

fredrikekelund commented on src/ipc-handlers.ts:
```diff
@@ -891,3 +893,23 @@ export async function openFileInIDE(
+		return;
+	}
+	await fsPromises.unlink( filePath );
+}
```
It looks like this code is currently unused. Should we add an IPC call to this in `onBackupCompleted`?

sejas replied:
Good catch! Now I'm calling it to remove it after we import a site: https://github.com/Automattic/studio/pull/612/commits/6baea65dc3ad493bc669b28c25cd2020faf4068d

fredrikekelund reviewed (commented):
Overall, this is looking great, and pulling works as expected 👍 I left a few code comments, though. Hopefully they are clear.

Also, a few comments about the functionality:
- There's currently no way to abort the pull operation. Should we address that in this PR?
- There's also no confirmation dialog.
- There are no inline success or error messages (see RToz6tIuQ7nlZrikBte4GU-fi-4491_115759)

I ran into an import error a few times during importing. They might have been flukes, but it would be interesting to see if any other team members can reproduce similar errors. The errors were possibly related to file permissions ( `Error: EPERM: operation not permitted, unlink`)

sejas committed:
Follow same convention for import and export site components

sejas committed:
block export while a site is being pulled

sejas committed:
Merge branch 'trunk' of github.com:Automattic/studio into add/sync-pull-remote-site

sejas committed:
Refactor context to live in its own folder and separate hooks

sejas committed:
Move the state management to the context to avoid using useSyncPull directly from a component by mistake

sejas committed:
extract setPullStates logic in favor of a custom callback

sejas committed:
do not update state before calling onBackupCompleted

sejas committed:
Disable start stop button when pulling a site

sejas committed:
capture sentry error when pulling

sejas committed:
capture error on sentry

sejas committed:
using timeout instead of interval as the state update will retrigger the effect

sejas committed:
Replace states for consistency, mapping old completed, current finished to downloading

sejas committed:
remove sync backup after imoprt

sejas committed:
add provider to tests

sejas committed:
add provider to remaining tests

sejas committed:
display pull complete state to allow user to clear it

sejas committed:
avoid jumping row height when changing states

sejas committed:
Display notificiation when pull completes making difference between staging and production

sejas committed:
Display error message inside connected sites list

sejas commented:
@katinthehatsite , @wojtekn ,
I've addressed all the feedback in this PR,  and it's in good shape for a second review.
I'll create an issue to add tests for the pull/push process.

katinthehatsite reviewed (commented):
Looking good!!

I think I see one small issue so far. This is what I observed:

* Connect a specific WPCOM site to Studio site A
* Connect **THAT SAME** WPCOM site to Studio site B
* Start the pull operation to Studio site A from the connected site
* Observe that the pull operation is in progress
* Switch to Studio site B
* Observe that pull operation is in progress for the connected site for which you started it on Studio site A

As long as I understand correctly, it seems that the pull operation should be in progress just on site A even though the same WPCOM site is connected to both site A and site B. Is that right? Or perhaps I might be missing something?

Finally, should we also ensure that the import tab does not go into `Import completed` state? What do you think?

<img width="678" alt="Screenshot 2024-11-13 at 6 31 57 PM" src="https://github.com/user-attachments/assets/5b8b4a2e-2270-4458-9f7c-9da764512ca2">

katinthehatsite commented:
One more thing 😅 

I noticed that when I disconnect a site with the pull completed and connect it back from the modal, the state with `Pull complete` and `Clear` persists:

<img width="693" alt="Screenshot 2024-11-13 at 6 36 30 PM" src="https://github.com/user-attachments/assets/0faf5713-b3d4-4442-93e1-53dbd039d755">

sejas committed:
clear import state after pull site to avoid showing import complete state on drop zone

sejas committed:
rename type and make sure we track the sync pull progress per local site

sejas committed:
refactor to use local site id and remote site id as pull state key

sejas committed:
create getPullState function to access the pull state

sejas committed:
update clear site key, and make sure that we clear pull state after disconnecting section

sejas commented:
@katinthehatsite, thanks for reviewing the PR again. I've updated the key to be a combination of local site id and remote site id, that way we won't share https://github.com/Automattic/studio/pull/612/commits/6f045ba3e23c2daf41f08d8e27900f305715babf

Now I'm clearing the pull state when a section is disconnected: https://github.com/Automattic/studio/pull/612/commits/6d7c92ed26e49db39b25d2075a092c2cb6688038

Here I'm clearing the import complete state after a pull successfully finishes. I think it's not a perfect solution in this case, but we can iterate on that later if we want: https://github.com/Automattic/studio/pull/612/commits/75753503e00becb173de70102959700a0f3cca9c
